### DISCUSSION TITLE
Threading / Messaging

### DISCUSSION THREAD

**Discussion by EdNekebno** [EdNekebno/DGTCentaurMods#40] [Ideas]
Just putting this here as we touched upon it a bit in a pull request. In my head a little bit the future direction of the boardfunctions code ideally goes:

* All of the existing functions to control the board (in terms of controlling/talking to the board) stay the same
* I want to thread the screen control code. So that everything using it can just write into an image which is a virtual screen buffer and it takes care of it (as working with the screen currently is so troublesome/difficult)
* Where a message comes from the board - key presses, piece lifts, etc I want to implement some kind of messaging/queue system. For this I'll take the eboard code when I know it is good to process and simplify the piece lift and put downs, etc. Then move detection is easy. Ideally a script could register for some kind of callback to know a key is pressed or piece lifted or put down or moved, etc. But I might need to improve my python skills to figure out how to do that. But essentially boardfunctions will do the heavy lifting

Though I have no set timeline for it. I'll do some research but if it'll take a while then I'll probably improve the lichess code first and if the Pegasus gets launched look at possibly emulating that. Then this. The screen I might take a look at earlier though (it wastes so much time debugging things!)

**Comment by wormstein** [comment]
Nice roadmap here.
What do you think starting with renaming boardfunctions to something shorter. Maybe just ''Board' or dgt? I know it comes with a lot of changes that could break trhings but with a good IDE is easy....

**Comment by EdNekebno** [comment]
Actually I couldn't resist prototyping a test for the screen idea. It works well. The only thing is it takes the epaper 1.2 seconds to refresh the display :(

I don't have a problem with boardfunctions.py being renamed to board.py .

**Reply by wormstein** [reply]
If yyou just feel the need to share the code with uz to get out opinion, just take the advantage of branching in git...

**Reply by EdNekebno** [reply]
If only I knew what I was doing in git :)

I actually went a bit further than I thought I would. It turns out it's not too tough. At the moment I have code that you can call a function in boardfunctions.py and it subscribes to keypresses. Also the screen code seems to work. Then I started working on a menu2.py using that functionality. It's still slow because of epaper but it makes submenus and things respond better/quicker. 

If I can do that then going forward we can use that new way of doing things. Lots of the old code will still be needed until things like lichess.py and eboard.py are migrated to it. But it's cool. You basically do:

`def keyPressed(id):
     # function receives key presses

def pieceMove(id):
     # function will receive piece lifts/placements

boardfunctions.clearSerial()
epaper.initEpaper()
boardfunctions.subscribeEvents(keyPressed, pieceMove)
time.sleep(5)
`

and you start to get events and can write to the screen.

**Comment by wormstein** [comment]
This is related to threading, so I put this here.

I was looking over the menu and to try see why is reacting so slow.
I think we have now to many while loops in there that is known as a bad way to halt execution. We should move to events based triggers. We should use <evebt>.set() and then use <event>.wait() where we need to listen for something.

This is not much of a diffference of your current way with a separate thread to poll events, but using set-wait will enable us to get rid of the while true loops...

Below I tried to build an example code to see what I mean:
``` python
#!/usr/bin/python3.7

import serial
import threading
import time

ser = serial.Serial("/dev/ttyS0", baudrate=1000000, timeout=0.2)

# Define events here
event_move = threading.Event()
event_key = threading.Event()

def main_loop():
    ping = bytearray(b'\x83\x06\x50\x59')
    reply = bytearray(b'\x85\x00\x06\x06\x50\x61')

    keycheck = bytearray(b'\x94\x06\x50\x6a')
    nokey = bytearray(b'\xb1\x00\x06\x06\x50\x0d')

    BTNUP = "b10011065000140a0508000000007d3c7c"

    while True:
        ser.write(ping)
        print(ping)
        resp = ser.read(10000)
        resp = bytearray(resp)
        print(resp)
        if (bytearray(resp) != reply):
            print("Nove event")
            move_event.set()
        time.sleep(0.1)

        ser.write(keycheck)
        print(keycheck)
        resp = bytearray(resp)
        resp = ser.read(10000)
        print(resp)
        if (bytearray(resp) != nokey):
            print("Key event")
            event_key.set()

        time.sleep(0.1)

if __name__ == "__main__":
    global key_pressed
    poll = threading.Thread(target=main_loop, args=(),daemon=True)
    poll.start()

    # We need a key, let's wait
    event_key.wait()
    print("Button is: " + key_pressed)

    # Have another function here to interpret the event.
```

**Comment by EdNekebno** [comment]
This kinda comes from my lack of understanding python and the way it does threading I suspect. I've done some reading. In principle this is good, in practice I'm not sure how much it'll help. There's an obvious issue where it's not so easy to wait for both events at once. e.g. if I want to wait for event_move and event_key at the same time in your example. This is easiest side-stepped I think by using data classes to set up some kind of message structure and just having one event that the board has sent a message. Then let the program decode which type it is.

The problem is if I look at eventsThread in the new boardfunctions code and integration in menu.py - it directly calls the function. Which should be automatically quicker. We can't actually use the loop around collecting the serial data this way as there must be something to trigger the original event. Although the time.sleep does seem long there so I'll change that. But there's no reduced time.sleeps to be had there. A single event could get rid of the time.sleep in the doMenu function.

For those things, this is just a few lines of code. So I'll do them as a test and see if it the menu feels any different

**Comment by wormstein** [comment]
It's amazing how fast centaur's menu reacts... Neither my python experience is not that good to figure out how they do it. I tried to look around with pyrasite and I see some objects named "event...".

**Comment by EdNekebno** [comment]
I pushed some changes (and broke it temporarily because I forgot to clear the event :) ) .Some experiments....

1. I reduced the time in eventsThread and added the event to doMenu. It doesn't feel much difference
2. CPU usage is low - it peaks at 30% when using the screen
3. If I print the time at which it first detects the keydown button to the end of the keyPressed then I get 1634568017.1612575  1634568017.214707.
4. From the boardfunctions tick key to doMenu completion: 1634568573.468835    1634568573.5216315

Then I'm with this in the sense we should do it where we can. But I don't think it'll have a noticeable time effect. I think responsiveness really is the screen refresh rate issue. If you click a button like the down button the board beeps almost straight away. It's just the screen is slow to show it. There's a way to update just some rows on the screen on many epaper displays - but I coded it in and it didn't work and took the same time as a full screen refresh. I'm convinced that is what the centaur process is doing (it's just updating the two menu lines - the one last selected and the one currently selected). If we could get that to work we can simply make epaper.py work out which rows it needs to refresh, crop it out of the buffer image, and it would be much quicker.

**Comment by wormstein** [comment]
Good idea to measure the time this way. It pinpoints clearly where the bottleneck is... And you're right about the beep, which is almost instant.

**Comment by EdNekebno** [comment]
There's a later display driver version. I'll see if it helps if I patch that in.

**Comment by EdNekebno** [comment]
New driver didn't help. However, I got partial updates to work sort off. I've pushed an update but not turned it on on the menu. If in menu.py you change:

epaper.initEpaper()   to   epaper.initEpaper(1)

it will enable it. You can see the speed difference. But also the problems. e.g. go into the Lichess menu. I'm not sure why it does this here but not in settings. Also a few more oddities that I think are related. Feel free to see if you can track them down.
